docs_src/design/ucr/Record-and-Replay.md

```diff
@@ -0,0 +1,54 @@
+## Record and Replay
+### Submitters
+- Lenny Goodell (Intel)
+- Jim White (IOTech)
+
+## Change Log
+- [pending](URL of PR) (2022-08-11)
+
+### Market Segments
+- Any/All
+
+### Motivation
+Currently one must have physical devices and appropriate environment to produce real device data (Event/Readings) into an EdgeX solution for other EdgeX services (Core Data, App Services, eKuiper Rules Engine) to consume. This is often not the case when someone is developing/testing one of these consuming EdgeX services. A good example of this is the **[RFID LLRP Inventory App Service](https://github.com/edgexfoundry/app-rfid-llrp-inventory)** . In order to test this service it is dependent on the [RFID LLRP Device Service](https://github.com/edgexfoundry/device-rfid-llrp-go), physical LLRP RFID readers, RFID Tags and environment where these are deployed. Having a way to record EdgeX the Event/Readings from an actual deployment that then could be replayed in development environment for testing would be very valuable. 
+
+Other potential uses are:
+
+- “Replayable data” could be used to validate a new or updated EdgeX service behaves like the existing one (EdgeX certification). Or the data could be offered by component providers to give users an example of the type of data generated from one of their device services.
+- EdgeX Ready applicants could use a record feature to capture data from their device service and the committee could explore that in relation to their profile to make sure they seemed to know what they were doing.
+- Record and replay can be used to reproduce a reported bug.
+- Record and export could serve as a simple means to backup data in the event of data loss.
+
+### Target Users
+- Device Manufacturer
+- Device Owner
+- Device User
+- Device Maintainer
+- Software Developer
+- Software Deployer
+- Software Integrator
+
+### Description
+Target users have the need to be able to replay recorded EdgeX Event/Readings for functional, performance or reproducible testing. This UCR describes a new capability that allows user to first record Event/Readings from real devices in real-time and than be able to replay the Event/Readings as if it was in real-time. The static device profile and device definition files at time of capture will need to be available and loaded at the time the captured data is replayed. 
+
+### Existing solutions
+There are simulators for some devices (i.e. Modbus), but there isn't an general solution to reproduce real device data into EdgeX without the physical devices being present. These simulators also don't have a way to produce a specific set of results in a timeline as do physical devices.
+
+### Requirements
+- Record device Event/Reading data (at rates, volumes and details like timestamps) 
```

Comment:
This requirement is an oversimplification of the following statement from the description:
> This UCR describes a new capability that allows user to first record Event/Readings from real devices in real-time and then be able to replay the Event/Readings as if it was in real-time.

I suggest expanding it to specify what exactly is needed in terms of data collection rate (real-time, at specific rate/frequency), period (how long to collect), amount (how much to collect e.g. total reading/event or bytes), details (what needs to be inside collected readings)

Reply:
> needed in terms of data collection rate (real-time, at specific rate/frequency)

The rate of Events is determined by the Devices and Device Services, so it will be was it is when received by the Recorder. The Event timestamps will suffice for determining the rate at time of capture. I will clarify in the requirement.

> period (how long to collect)
Yes, this should be spelled out in the requirements

> amount (how much to collect e.g. total reading/event or bytes)
I was thinking Period would suffice, but if ya have infrequent Readings you may want it to stop after a set number of Readings. I will add to the requirements

Reply:
Restructured and added